### DISCUSSION TITLE
firebase_storage upload in background

### DISCUSSION THREAD

**Discussion by syonip** [firebase/flutterfire#5610] [Feature request]
**Is your feature request related to a problem? Please describe.**
Uploading large files (such as video files) could take a long time. Right now if the app is put in the background while an upload is running, it can close and stop the upload.

**Describe the solution you'd like**
I would like the upload process to continue running when the app is in the background / closed.

**Comment by techno-disaster** [comment]
@syonip were you able to find a workaround? I'm planning to use [flutter_uploader](https://pub.dev/packages/flutter_uploader) with this. 

cc : @TahaTesser

**Comment by syonip** [comment]
@Techno-Disaster Yeah I'm just writing a blog post about a workaround using cloud functions to generate an upload url.
Full guide should be done soon, here is the gist of it in the meanwhile:
![flutter_fbstorage_background_upload](https://user-images.githubusercontent.com/5416329/97833558-5045c800-1cde-11eb-909d-803d2466bfe6.jpg)
 
Edit: you can read the full guide here: https://medium.com/@syonip/background-video-upload-to-firebase-storage-with-flutter-ee398b796656

Source: https://github.com/syonip/flutter_fbstorage_background_upload

**Comment by techno-disaster** [comment]
> @Techno-Disaster Yeah I'm just writing a blog post about a workaround using cloud functions to generate an upload url.
> Full guide should be done soon, here is the gist of it in the meanwhile:
> ![flutter_fbstorage_background_upload](https://user-images.githubusercontent.com/5416329/97833558-5045c800-1cde-11eb-909d-803d2466bfe6.jpg)
> 
> Edit: you can read the full guide here: https://medium.com/@syonip/background-video-upload-to-firebase-storage-with-flutter-ee398b796656
> 
> Source: https://github.com/syonip/flutter_fbstorage_background_upload

ah thanks a lot man, I'm implementing something similar for images this is really helpful 💙

**Comment by rlee1990** [comment]
@TahaTesser any updates on this?

**Comment by jamesdixon** [comment]
@syonip nice workaround! My only concern with this is the additional cost incurred by having to use Cloud Functions to get the resumable URL.

@TahaTesser is there any update on this? This is actually a pretty big deal. We have an application where a user may upload hundreds of photos in a given session and are often in areas where cell service is not great. Forcing them to go back into the app to wait for all photos to upload is problematic.

**Reply by jamesdixon** [reply]
@syonip does your solution work if the user happens to be offline/bad service when they initiate the upload? I'm guessing not given the reliance on cloud functions but figured I'd ask in case I missed something :)

**Reply by syonip** [reply]
@jamesdixon I think the cost of one cloud function per-upload is not significant compared to the storage costs, you get 2M function calls per month just in the free plan :)

About offline, the background upload won't start before you get the upload url, so if the app goes into background before you go back online, it won't start. To support that use-case you'd have to have some retry mechanism running in a foreground notification on android, but on iOS it's probably not possible because of iOS limitations.
I think most apps don't support this kind of thing though.

**Reply by techno-disaster** [reply]
iirc, even if connection breaks and the app is in recents, eventually the upload is retried. Also when incase you close the app and reopen it any failed uploads in queue are retried. (you might want to test this last point tho, not sure)
